server/market/market.go

```diff
@@ -1183,30 +1184,27 @@ func (m *Market) collectPreimages(orders []order.Order) (cSum []byte, ordersReve
 			preimage: piChan,
 		}
 
-		// Failure to respond in time is a miss, signalled by a nil pointer.
-		var missOnce sync.Once // captured by miss
-		miss := func() {
-			// RequestWithTimeout should only call this on expire, not because
-			// of link or other errors, but put it in a sync.Once to be safe
-			// since it will block if run twice.
-			missOnce.Do(func() { piChan <- nil })
-		}
+		// Failure to respond in time or an async link write error is a miss,
+		// signalled by a nil pointer. Request errors returned by
+		// RequestWithTimeout instead register a miss immediately.
+		miss := func() { piChan <- nil }
 
 		// Send the preimage request to the order's owner.
 		err = m.auth.RequestWithTimeout(ord.User(), req, func(_ comms.Link, msg *msgjson.Message) {
-			m.handlePreimageResp(msg, reqData)
+			m.handlePreimageResp(msg, reqData) // sends on piChan
 		}, piTimeout, miss)
 		if err != nil {
-			miss() // only called by RequestWithTimeout on expire, not send errors
-			if errors.Is(err, ws.ErrPeerDisconnected) {
-				misses = append(misses, ord)
```

Comment:
The `miss()` call here was superfluous since there was a `continue` below skipping the `preimages[ord] = piChan`, thus preventing the subsequent `range preimages` loop from waiting on a result from this order.  It was just sending nil into a buffered channel that nothing ever received from.

The substantive change here is just doing `misses = append(misses, ord)` for any non-nil `err` before continuing to the next order's preimage request.